[CI] GlobDiscoveryLocalTests testRecursiveDoubleStarAllFiles failing

**Build Scans:**
- [elasticsearch-periodic #11893 / single-processor-node-tests](https://gradle-enterprise.elastic.co/s/vqea7qkghisxg)
- [elasticsearch-periodic-platform-support #11870 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/hiy2hohbjmfcs)
- [elasticsearch-periodic-platform-support #11870 / amazonlinux-2023_platform-support-aws](https://gradle-enterprise.elastic.co/s/epbt3fber24y4)
- [elasticsearch-periodic-platform-support #11870 / windows-2022_checkpart3_platform-support-windows](https://gradle-enterprise.elastic.co/s/rhntfz3tl3ly4)
- [elasticsearch-periodic-platform-support #11870 / windows-2025_checkpart3_platform-support-windows](https://gradle-enterprise.elastic.co/s/aync6p6czmql2)
- [elasticsearch-pull-request #121355 / part-3](https://gradle-enterprise.elastic.co/s/jjgizg2unujik)
- [elasticsearch-pull-request #121331 / part-3](https://gradle-enterprise.elastic.co/s/il6jbmdrh7jsw)
- [elasticsearch-pull-request #121295 / part-3](https://gradle-enterprise.elastic.co/s/wlmoz6lljp5ay)
- [elasticsearch-pull-request #121280 / part-3](https://gradle-enterprise.elastic.co/s/vbkubb6xzmzdc)
- [elasticsearch-pull-request #121275 / part-3](https://gradle-enterprise.elastic.co/s/3kkkymmxtzzvu)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:esql:test" --tests "org.elasticsearch.xpack.esql.datasources.GlobDiscoveryLocalTests.testRecursiveDoubleStarAllFiles" -Dtests.seed=8D660F892A7C26F1 -Dtests.configure_test_clusters_with_one_processor=true -Dtests.locale=fi-Latn-FI -Dtests.timezone=America/Scoresbysund -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.esql.datasources.GlobDiscoveryLocalTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testRecursiveDoubleStarAllFiles'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected:<5> but was:<8>
```

**Issue Reasons:**
- [main] 2 consecutive failures in step amazonlinux-2023_platform-support-aws
- [main] 17 failures in test testRecursiveDoubleStarAllFiles (12.5% fail rate in 136 executions)
- [main] 2 failures in step amazonlinux-2023_platform-support-aws (100.0% fail rate in 2 executions)
- [main] 12 failures in step part-3 (14.1% fail rate in 85 executions)
- [main] 12 failures in pipeline elasticsearch-pull-request (14.6% fail rate in 82 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -219,9 +219,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.qa.multi_node.EsqlClientYamlIT
   method: test {p0=esql/160_union_types/suggested_type}
   issue: https://github.com/elastic/elasticsearch/issues/142525
-- class: org.elasticsearch.xpack.esql.datasources.GlobDiscoveryLocalTests
-  method: testRecursiveDoubleStarAllFiles
-  issue: https://github.com/elastic/elasticsearch/issues/142576
 - class: org.elasticsearch.xpack.esql.qa.mixed.EsqlClientYamlIT
   method: test {p0=esql/160_union_types/suggested_type}
   issue: https://github.com/elastic/elasticsearch/issues/142525

--- a/x-pack/plugin/esql-datasource-http/src/main/java/org/elasticsearch/xpack/esql/datasource/http/local/LocalStorageProvider.java
+++ b/x-pack/plugin/esql-datasource-http/src/main/java/org/elasticsearch/xpack/esql/datasource/http/local/LocalStorageProvider.java
@@ -118,11 +118,7 @@ public final class LocalStorageProvider implements StorageProvider {
      */
     @SuppressForbidden(reason = "LocalStorageProvider converts user-supplied file:// URIs to Path objects")
     private Path toFilePath(StoragePath storagePath) {
-        String pathStr = storagePath.path();
-
-        // Handle file:// URLs - the path() method returns the path component after the scheme
-        // For file:///absolute/path, path() returns "/absolute/path"
-        // For file://relative/path, path() returns "relative/path"
+        String pathStr = storagePath.localPath();
 
         if (pathStr == null || pathStr.isEmpty()) {
             throw new IllegalArgumentException("Path cannot be empty for file:// scheme");

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/datasources/spi/StoragePath.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/datasources/spi/StoragePath.java
@@ -138,6 +138,19 @@ public final class StoragePath {
         return path;
     }
 
+    /**
+     * Returns the path component adjusted for use as a local filesystem path.
+     * On Windows, file:// URIs produce paths like {@code /C:/dir/file} where the
+     * leading slash is invalid for the OS. This method strips it when a drive letter
+     * is detected so the result can be passed to {@code PathUtils.get()} safely.
+     */
+    public String localPath() {
+        if (path.length() >= 3 && path.charAt(0) == '/' && Character.isLetter(path.charAt(1)) && path.charAt(2) == ':') {
+            return path.substring(1);
+        }
+        return path;
+    }
+
     public String objectName() {
         if (path.isEmpty() || path.equals(PATH_SEPARATOR)) {
             return "";

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/datasources/GlobDiscoveryLocalTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/datasources/GlobDiscoveryLocalTests.java
@@ -281,7 +281,7 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
 
         @Override
         public StorageIterator listObjects(StoragePath prefix, boolean recursive) throws IOException {
-            Path dirPath = PathUtils.get(prefix.path());
+            Path dirPath = PathUtils.get(prefix.localPath());
             if (Files.exists(dirPath) == false || Files.isDirectory(dirPath) == false) {
                 return emptyIterator();
             }
@@ -291,7 +291,7 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
                 Files.walkFileTree(dirPath, new SimpleFileVisitor<>() {
                     @Override
                     public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
-                        if (attrs.isRegularFile()) {
+                        if (attrs.isRegularFile() && isExtraFSFile(file) == false) {
                             entries.add(toEntry(file, attrs));
                         }
                         return FileVisitResult.CONTINUE;
@@ -301,7 +301,7 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
                 try (DirectoryStream<Path> stream = Files.newDirectoryStream(dirPath)) {
                     for (Path entry : stream) {
                         BasicFileAttributes attrs = Files.readAttributes(entry, BasicFileAttributes.class);
-                        if (attrs.isRegularFile()) {
+                        if (attrs.isRegularFile() && isExtraFSFile(entry) == false) {
                             entries.add(toEntry(entry, attrs));
                         }
                     }
@@ -330,7 +330,7 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
 
         @Override
         public boolean exists(StoragePath path) {
-            return Files.exists(PathUtils.get(path.path()));
+            return Files.exists(PathUtils.get(path.localPath()));
         }
 
         @Override
@@ -344,6 +344,10 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
         private static StorageEntry toEntry(Path file, BasicFileAttributes attrs) {
             StoragePath storagePath = StoragePath.of(StoragePath.fileUri(file));
             return new StorageEntry(storagePath, attrs.size(), attrs.lastModifiedTime().toInstant());
+        }
+
+        private static boolean isExtraFSFile(Path file) {
+            return file.getFileName().toString().startsWith("extra");
         }
 
         private static StorageIterator emptyIterator() {
@@ -393,7 +397,7 @@ public class GlobDiscoveryLocalTests extends ESTestCase {
 
         @Override
         public boolean exists() {
-            return Files.exists(PathUtils.get(path.path()));
+            return Files.exists(PathUtils.get(path.localPath()));
         }
 
         @Override

